### PR TITLE
lorawan: update LoRaMac-node library

### PR DESCRIPTION
**Pull request by tagunil**
Update the LoRaMac-node library to the last stable release and fix the Zephyr glue code to match it.
Depends on https://github.com/zephyrproject-rtos/loramac-node/pull/11

### CODE REVIEW

**Comment by tagunil** [comment]
Well, the way we're building this library is quite messy. Looks like we build the entire LoRaMac stack every time, even if all we want is just a radio driver. Naturally this causes build issues here and there.

**Comment by tagunil** [comment]
@Mani-Sadhasivam I've rearranged some Kconfig symbol dependencies to represent actual LoRaMac-node code dependencies. Could you please look at these? Here and in https://github.com/zephyrproject-rtos/loramac-node/pull/9

File: drivers/lora/Kconfig.sx12xx [hunk 1]
```diff
@@ -12,7 +12,6 @@ DT_COMPAT_SEMTECH_SX1262 := semtech,sx1262
 menuconfig LORA_SX12XX
 	bool "Semtech SX-series driver"
 	select HAS_SEMTECH_RADIO_DRIVERS
-	select HAS_SEMTECH_LORAMAC
```
Comment (JordanYates) [inline drivers/lora/Kconfig.sx12xx:15]: Splitting the Kconfig changes unrelated to the update into a separate commit would be nice.

File: drivers/lora/sx126x.c [hunk 2]
```diff
@@ -492,7 +501,7 @@ static int sx126x_lora_init(const struct device *dev)
 	dev_data.spi = device_get_binding(DT_INST_BUS_LABEL(0));
 	if (!dev_data.spi) {
 		LOG_ERR("Cannot get pointer to %s device",
-			    DT_INST_BUS_LABEL(0));
```
Comment (JordanYates) [inline drivers/lora/sx126x.c:495]: Unrelated whitespace changes

File: drivers/lora/sx126x.c [hunk 3]
```diff
@@ -531,5 +540,5 @@ static const struct lora_driver_api sx126x_lora_api = {
 };
 
 DEVICE_DT_INST_DEFINE(0, &sx126x_lora_init, NULL, NULL,
-		    NULL, POST_KERNEL, CONFIG_LORA_INIT_PRIORITY,
-		    &sx126x_lora_api);
+		      NULL, POST_KERNEL, CONFIG_LORA_INIT_PRIORITY,
```
Comment (JordanYates) [inline drivers/lora/sx126x.c:543]: Unrelated whitespace changes

File: drivers/lora/sx126x.c [hunk 4]
```diff
@@ -443,6 +443,15 @@ void SX126xWakeup(void)
 	dev_data.mode = MODE_STDBY_RC;
 }
 
+uint32_t SX126xGetDio1PinState(void)
+{
+	if (gpio_pin_get(dev_data.dio1, GPIO_DIO1_PIN) > 0) {
```
Comment (JordanYates) [inline drivers/lora/sx126x.c:448]: I'm a fan of ternary statements in situations like this:
```
return gpio_pin_get(dev_data.dio1, GPIO_DIO1_PIN) > 0 ? 1U : 0U;
```
Just a suggestion.

File: drivers/lora/sx127x.c [hunk 5]
```diff
@@ -511,6 +516,17 @@ void SX127xSetRfTxPower(int8_t power)
 	}
 }
 
+uint32_t SX127xGetDio1PinState(void)
+{
+#if SX127X_DIO_GPIO_LEN(0) >= 2
```
Comment (JordanYates) [inline drivers/lora/sx127x.c:521]: Is this necessary? Can the SX127x drivers operate without DIO1?
The loramac-node drivers will certainly error out in `SX127xIoIrqInit` if DIO1 doesn't exist.

File: drivers/lora/Kconfig.sx12xx (hunk 1 above)
Comment (tagunil) [reply drivers/lora/Kconfig.sx12xx:15]: I thought so, but these Kconfig changes must be synchronized with CMakeLists.txt changes in loramac-node module. Overall they are required to avoid build errors.

File: drivers/lora/sx126x.c (hunk 2 above)
Comment (tagunil) [reply drivers/lora/sx126x.c:495]: Ah, yes. I'll split these into a separate commit.

File: drivers/lora/sx127x.c (hunk 5 above)
Comment (tagunil) [reply drivers/lora/sx127x.c:521]: Well, I'm not sure about operating without DIO1, but I've tried to make functional changes as minimal as possible. This is a library update, after all.

File: drivers/lora/Kconfig.sx12xx (hunk 1 above)
Comment (tagunil) [reply drivers/lora/Kconfig.sx12xx:15]: I just don't like unbuildable commits.

File: drivers/lora/sx127x.c (hunk 5 above)
Comment (JordanYates) [reply drivers/lora/sx127x.c:521]: Yes, but your change to to sx1262x doesn't do any checking that DIO1 exists.
Why does the sx127x driver need it, or why doesn't the sx1262 need it?

File: drivers/lora/Kconfig.sx12xx (hunk 1 above)
Comment (JordanYates) [reply drivers/lora/Kconfig.sx12xx:15]: Thats fine, it wasn't obvious that these changes aren't just cosmetic, and I'm not going through a 375 commit diff to find that out.

File: drivers/lora/sx127x.c (hunk 5 above)
Comment (tagunil) [reply drivers/lora/sx127x.c:521]: These drivers are different. For the sx126x driver DIO1 is a hard requirement, the sx127x doesn't require it. I don't think it is good; in my opinion, both drivers need refactoring and more unification. But they work that way at the moment.

File: drivers/lora/sx126x.c (hunk 4 above)
Comment (tagunil) [reply drivers/lora/sx126x.c]: Done.

File: drivers/lora/sx126x.c (hunk 2 above)
Comment (tagunil) [reply drivers/lora/sx126x.c:495]: Fixed.

File: drivers/lora/sx126x.c (hunk 3 above)
Comment (tagunil) [reply drivers/lora/sx126x.c:539]: Fixed.

**Review by Mani-Sadhasivam — APPROVED** [review]
LGTM!

**Comment by JordanYates** [comment]
This should not be merged as is as it conflicts with the decision made here: https://github.com/zephyrproject-rtos/loramac-node/pull/10#issuecomment-859986651

**Comment by tagunil** [comment]
@JordanYates, @Mani-Sadhasivam, I've adapted this PR to our new structure.

**Review by Mani-Sadhasivam — APPROVED** [review]

**Comment by Mani-Sadhasivam** [comment]
@tagunil Please resolve the conflict

**Comment by tagunil** [comment]
> @tagunil Please resolve the conflict

Done.

**Comment by Mani-Sadhasivam** [comment]
@mniestroj @JordanYates Please review!

File: modules/loramac-node/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX1272 ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx1272/sx1272.c)
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX1276 ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx1276/sx1276.c)
+
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX126X ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx126x/sx126x.c)
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX126X ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx126x/radio.c)
+
+if(${CONFIG_HAS_SEMTECH_LORAMAC})
+zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/mac)
+zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/mac/region)
```
Comment (tejlmand) [inline modules/loramac-node/CMakeLists.txt:23]: please indent this two lines:
```suggestion
  zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/mac)
  zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/mac/region)
```

File: modules/loramac-node/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/system)
+zephyr_library_include_directories(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio)
+
+zephyr_library_sources(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/boards/mcu/utilities.c)
+
+zephyr_library_sources(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/system/timer.c)
+zephyr_library_sources(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/system/delay.c)
+zephyr_library_sources(${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/system/systime.c)
+
+zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX1272 ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx1272/sx1272.c)
```
Comment (tejlmand) [inline modules/loramac-node/CMakeLists.txt:15]: There are some very long lines here and below (>100 chars long).

Could you please wrap those long lines, for example as:
```
zephyr_library_sources_ifdef(CONFIG_HAS_SEMTECH_SX1272
  ${ZEPHYR_LORAMAC_NODE_MODULE_DIR}/src/radio/sx1272/sx1272.c
)
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Generally looking good, some minor nits observed.

**Review by tejlmand — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
Hey @tagunil, thanks for the patch, great to have the updated stack! Unfortunately though it looks like this had the side effect of disabling USE_RANDOM_DEV_NONCE (from https://github.com/zephyrproject-rtos/loramac-node/commit/ff05e9b6a16ee9aebbcee613532e1eb2f0df6716), meaning that the sample code (and anything that does not store the join counter) is only going to be able to connect once to network like TTN, or anything with unique DevNonce checks on the activation server.

Do you know if there's a workaround or work in progress to restore the DevNonce randomization functionality?

**Comment by tagunil** [comment]
> Hey @tagunil, thanks for the patch, great to have the updated stack! Unfortunately though it looks like this had the side effect of disabling USE_RANDOM_DEV_NONCE (from https://github.com/zephyrproject-rtos/loramac-node/commit/ff05e9b6a16ee9aebbcee613532e1eb2f0df6716), meaning that the sample code (and anything that does not store the join counter) is only going to be able to connect once to network like TTN, or anything with unique DevNonce checks on the activation server.
> 
> Do you know if there's a workaround or work in progress to restore the DevNonce randomization functionality?

As far as I know, monotonically incrementing DevNonce is mandated by the latest LoRaWAN specification, so we should rather provide the means to store all these counters in the non-volatile memory.

**Comment by tagunil** [comment]
@fabiobaltieri, could you please create an issue to track this requirement?

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri, could you please create an issue to track this requirement?

Sure thing!
